framework/src/mfem/solvers/MFEMCGSolver.C

```diff
@@ -40,8 +36,12 @@ MFEMCGSolver::constructSolver(const InputParameters &)
   solver->SetMaxIter(getParam<int>("l_max_its"));
   solver->SetPrintLevel(getParam<int>("print_level"));
 
-  if (_preconditioner)
+  if (isParamSetByUser("preconditioner"))
+  {
+    auto & pre_ref = const_cast<MFEMSolverBase &>(getUserObject<MFEMSolverBase>("preconditioner"));
+    _preconditioner = std::dynamic_pointer_cast<MFEMSolverBase>(pre_ref.getSharedPtr());
```

Comment:
Do we need `_preconditioner` to be a shared pointer? Would just a regular pointer suffice? Then we can avoid bouncing up and down the hierarchy. MFEMProblem will definitely destruct after this object

Reply:
I just kept it as it was originally, but I agree it seems unnecessary for `_preconditioner`, so that's now a raw pointer.
I've left `_solver` as a shared pointer though, as changing that in the same way isn't (unless I'm missing something) quite as straightforward, we'd need to manage that memory ourselves.